File: Java/sortedListToBST.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; next = null; }
 * }
 */
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        // IMPORTANT: Please reset any member data you declared, as
        // the same Solution instance will be reused for each test case.
        int len = getListLength(head);
        
        return sortedListToBSTHelper(head, 0, len - 1);
    }
    
    private TreeNode sortedListToBSTHelper(ListNode head, int begin, int end)
    {
        if (begin > end)
	    {
		return null;
	    }
        
        int mid = (end - begin) / 2 + begin;
        
        TreeNode left = sortedListToBSTHelper(head, begin, mid - 1);
        TreeNode root = new TreeNode(head.val);
        root.left = left;
        head = head.next;
        root.right = sortedListToBSTHelper(head, mid + 1, end);
        
        return root;
    }  
    
    private int getListLength(ListNode head)
    {
        if (head == null)
	    {
		return 0;
	    }
        
        int len = 0;
        ListNode current = head;
        while (current != null)
	    {
		current = current.next;
		len++;
	    }
        
        return len;
    }
}

File: C++/sortedListToBST.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        ListNode *ptr = head;
        int length = 0;
        
        while (ptr != NULL)
        {
            ptr = ptr->next;
            length++;
        }
        
        return listToBSTHelper(head, 0, length - 1);
    }
    
    TreeNode *listToBSTHelper(ListNode *&head, int low, int high)
    {
        if (low > high)
        {
            return NULL;
        }
        
        int mid = (high - low) / 2 + low;
        
        TreeNode *lchild = listToBSTHelper(head, low, mid - 1);
        TreeNode *root = new TreeNode(head->val);
        root->left = lchild;
        head = head->next;
        root->right = listToBSTHelper(head, mid + 1, high);
        
        return root;         
    }
};